### PR TITLE
Upstream/rt5677

### PR DESCRIPTION
**Pull request by cujomalainey**
Objectives:
Update to new alsa API and squash into original changes
Squash out as many intermediate changes as possible
Fix any bugs introduced into new API and squash changes

Known limitations:
Due to the I2S MCLK1 being hidden completely behind the bdw platform device and changes being introduced post 3.14 that disable the clock on supsned, there is no way to prevent it the clock from suspending before the codec in an acceptable abstracted fashion. The current workaround relies on userspace to the toggle the kcontrol before suspending the codec so the codec DSP shuts-down with the MCLK still active. If this doesn't happen, the writes are ignored as the codec is still in DSP mode and has no clock, therefore corrupting the regmap. In order to recover the state, the pcm needs to be closed and opened again in this state. Any input on this issue would be greatly appreciated.

### CODE REVIEW

File: sound/soc/codecs/rt5677-spi.c [hunk 1]
```diff
@@ -45,9 +47,340 @@
...
+	bool new_hotword;	/* a new hotword is fired */
+};
+
+static const struct snd_pcm_hardware rt5677_spi_pcm_hardware = {
+	.info			= SNDRV_PCM_INFO_MMAP |
+				  SNDRV_PCM_INFO_MMAP_VALID |
+				  SNDRV_PCM_INFO_INTERLEAVED,
+	.formats		= SNDRV_PCM_FMTBIT_S16_LE,
+	.period_bytes_min	= PAGE_SIZE,
+	.period_bytes_max	= 0x20000 / 8,
```
Comment (dgreid) [inline sound/soc/codecs/rt5677-spi.c]: Should 0x20000 be a constant defined above?

File: sound/soc/codecs/rt5677-spi.c [hunk 2]
```diff
@@ -45,9 +47,340 @@
...
+	.periods_min		= 8,
+	.periods_max		= 8,
+	.channels_min		= 1,
+	.channels_max		= 1,
+	.buffer_bytes_max	= 0x20000,
+};
+
+static int rt5677_spi_dai_hw_params(struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params, struct snd_soc_dai *dai)
+{
```
Comment (dgreid) [inline sound/soc/codecs/rt5677-spi.c]: Would a comment be helpful about why this empty function is needed?

File: sound/soc/codecs/rt5677.c [hunk 3]
```diff
@@ -710,74 +918,52 @@ static int rt5677_set_dsp_vad(struct snd_soc_component *component, bool on)
 	if (!IS_ENABLED(CONFIG_SND_SOC_RT5677_SPI))
 		return -ENXIO;
 
+	dev_info(component->dev, "DSP VAD: on=%d, activity=%d\n", on, activity);
 	if (on && !activity) {
 		activity = true;
 
-		regcache_cache_only(rt5677->regmap, false);
```
Comment (dgreid) [inline sound/soc/codecs/rt5677.c:714]: Were these cache controls not needed anyways?

**Review by dgreid — COMMENTED** [review]
overall seems good to me, but I just skimmed most of it.

File: sound/soc/codecs/rt5677-spi.c (hunk 1 above)
Comment (cujomalainey) [reply sound/soc/codecs/rt5677-spi.c]: I will have to talk to ben about this, I am guessing this comes from the firmware.

File: sound/soc/codecs/rt5677-spi.c (hunk 2 above)
Comment (cujomalainey) [reply sound/soc/codecs/rt5677-spi.c]: Most likely not needed. Thanks for catching this.

File: sound/soc/codecs/rt5677.c (hunk 3 above)
Comment (cujomalainey) [reply sound/soc/codecs/rt5677.c:714]: I will track down the history and quash out if needed

File: sound/soc/codecs/rt5677-spi.c (hunk 2 above)
Comment (cujomalainey) [reply sound/soc/codecs/rt5677-spi.c]: deleted

**Pull request closed without merging**